src/Go/boardAnalysis/boardAnalysis.ts

```diff
@@ -0,0 +1,519 @@
+import {
+  Board,
+  BoardState,
+  Neighbor,
+  opponents,
+  PlayerColor,
+  playerColors,
+  PointState,
+  validityReason,
+} from "../boardState/goConstants";
+import {
+  findAdjacentPointsInChain,
+  findNeighbors,
+  getArrayFromNeighbor,
+  getBoardCopy,
+  getEmptySpaces,
+  getNewBoardState,
+  getStateCopy,
+  isDefined,
+  isNotNull,
+  updateCaptures,
+  updateChains,
+} from "../boardState/boardState";
+
+/**
+ * Determines if the given player can legally make a move at the specified coordinates.
+ *
+ * You cannot repeat previous board states, to prevent endless loops (superko rule)
+ *
+ * You cannot make a move that would remove all liberties of your own piece(s) unless it captures opponent's pieces
+ *
+ * You cannot make a move in an occupied space
+ *
+ * You cannot make a move if it is not your turn, or if the game is over
+ *
+ * @returns a validity explanation for if the move is legal or not
+ */
+export function evaluateIfMoveIsValid(boardState: BoardState, x: number, y: number, player: PlayerColor) {
+  const point = boardState.board?.[x]?.[y];
```

Comment:
Ah, nvm. I see now that you can't trust the input as it could be coming in from scripts and untrusted input. I wonder if there should be a more specific error for invalid input.

Reply:
It can't reasonably be empty, but typescript had strong opinions about how this should be written.